File: backend/app/schemas/consommation.py
"""Schémas Pydantic pour SessionConsommation"""
from datetime import datetime
from typing import Any, Dict, Optional
from pydantic import BaseModel, Field


class SessionConsommationCreate(BaseModel):
    date_heure:       Optional[datetime]   = None   # None → datetime.utcnow côté serveur
    id_vaporisateur:  Optional[int]        = None
    type_produit:     str                           # fleur | hash | rosin
    id_stock:         Optional[int]        = None
    quantite_g:       float                = Field(..., gt=0)
    options_vapo:     Optional[Dict[str, Any]] = None
    notes:            Optional[str]        = None


class SessionConsommationUpdate(BaseModel):
    date_heure:       Optional[datetime]   = None
    id_vaporisateur:  Optional[int]        = None
    type_produit:     Optional[str]        = None
    id_stock:         Optional[int]        = None
    quantite_g:       Optional[float]      = None
    options_vapo:     Optional[Dict[str, Any]] = None
    notes:            Optional[str]        = None


class SessionConsommationRead(BaseModel):
    id_session:       int
    date_heure:       datetime
    id_vaporisateur:  Optional[int]
    nom_vaporisateur: Optional[str]        = None   # enrichi
    type_produit:     str
    id_stock:         Optional[int]
    nom_variete:      Optional[str]        = None   # enrichi depuis Stock → Variete
    quantite_g:       float
    options_vapo:     Optional[Dict[str, Any]]
    notes:            Optional[str]
    created_at:       Optional[datetime]

    class Config:
        from_attributes = True

File: frontend/src/api/consommation.ts
import client from './client'

export interface SessionConsommation {
  id_session:       number
  date_heure:       string
  id_vaporisateur?: number | null
  nom_vaporisateur?: string | null
  type_produit:     string   // fleur | hash | rosin
  id_stock?:        number | null
  nom_variete?:     string | null
  quantite_g:       number
  options_vapo?:    Record<string, unknown> | null
  notes?:           string | null
  created_at?:      string | null
}

export interface SessionConsommationCreate {
  date_heure?:      string
  id_vaporisateur?: number | null
  type_produit:     string
  id_stock?:        number | null
  quantite_g:       number
  options_vapo?:    Record<string, unknown> | null
  notes?:           string | null
}

export interface ConsoStats {
  periodes: { jour: number; semaine: number; mois: number; annee: number }
  by_type:  Record<string, number>
  by_vapo:  { id_vaporisateur: number | null; nom: string; total_g: number }[]
  last7:    { date: string; total_g: number }[]
  avg_7j_g: number
  stock_dispo_g: Record<string, number>
  jours_restants: number | null
}

export const consommationAPI = {
  getAll: (limit = 100, offset = 0) =>
    client.get<SessionConsommation[]>('/consommation/', { params: { limit, offset } }).then(r => r.data),

  getStats: () =>
    client.get<ConsoStats>('/consommation/stats').then(r => r.data),

  create: (data: SessionConsommationCreate) =>
    client.post<SessionConsommation>('/consommation/', data).then(r => r.data),

  update: (id: number, data: Partial<SessionConsommationCreate>) =>
    client.put<SessionConsommation>(`/consommation/${id}`, data).then(r => r.data),

  delete: (id: number) =>
    client.delete(`/consommation/${id}`),
}
